### PR TITLE
[GLUTEN-7110][VL][DELTA] support IncrementMetric in gluten

### PR DESCRIPTION
**Pull request by dcoliversun**
## What changes were proposed in this pull request?

This PR aims to support IncrementMetric in gluten. In delta, IncrementMetric is used:  
* DeleteCommand: Filter
  * https://github.com/delta-io/delta/blob/cdd39dda25f1a9b06ce7a414d0ebff4ddfa419c5/spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala#L311
  * https://github.com/delta-io/delta/blob/cdd39dda25f1a9b06ce7a414d0ebff4ddfa419c5/spark/src/main/scala/org/apache/spark/sql/delta/commands/DeleteCommand.scala#L461
 * UpdateCommand: Filter
   * https://github.com/delta-io/delta/blob/cdd39dda25f1a9b06ce7a414d0ebff4ddfa419c5/spark/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala#L188
   * https://github.com/delta-io/delta/blob/cdd39dda25f1a9b06ce7a414d0ebff4ddfa419c5/spark/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala#L366
 * MergeIntoCommand: Project
   * https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/commands/merge/ClassicMergeExecutor.scala

(Fixes: \#7110)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala [hunk 1]
```diff
@@ -73,9 +77,17 @@ abstract class FilterExecTransformerBase(val cond: Expression, val input: SparkP
       validation: Boolean): RelNode = {
     assert(condExpr != null)
     val args = context.registeredFunction
-    val condExprNode = ExpressionConverter
-      .replaceWithExpressionTransformer(condExpr, attributeSeq = originalInputAttributes)
-      .doTransform(args)
+    val condExprNode = condExpr match {
+      case IncrementMetric(child, metric) =>
```
Comment (yma11) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala:81]: Can we create a class similar as `CHFilterExecTransformer`, such `DeltaFilterExecTransformer` and override `getRelNode` method instead? We should put it under `gluten-delta` folder thus to keep core clean from data lake specific logic and no need add delta package as dependency for `gluten-substrait`.

Comment (dcoliversun) [reply gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala:81]: Sure. I will do it after Project supports IncrementMetric

File: gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.gluten.substrait.rel.{RelBuilder, RelNode}
+
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, Expression}
+import org.apache.spark.sql.delta.metric.IncrementMetric
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.metric.SQLMetric
+
+import scala.collection.JavaConverters._
+
+case class DeltaFilterExecTransformer(condition: Expression, child: SparkPlan)
```
Comment (yma11) [inline gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: Whether can we put more code in `FilterExecTransformerBase` if delta can share the logic?

Comment (dcoliversun) [reply gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: Sure, I am doing on this

Comment (dcoliversun) [reply gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: @yma11 Updated. And some Clickhouse unit tests failed in Jenkins. Could you help me to check exception information? I have no account.

Comment (yma11) [reply gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: [log.txt](https://github.com/user-attachments/files/16972027/log.txt)

Comment (dcoliversun) [reply gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: Thank you for your help. The ClickHouse unit tests are OK. I am currently conducting end-to-end testing and investigating the reason for a regression.  

image

Comment (dcoliversun) [reply gluten-delta/src/main/delta-32/org/apache/gluten/execution/DeltaFilterExecTransformer.scala:34]: Will be fixed in #7483

File: gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala [hunk 3]
```diff
@@ -160,7 +178,7 @@ object FilterExecTransformerBase {
   }
 }
 
-case class ProjectExecTransformer private (projectList: Seq[NamedExpression], child: SparkPlan)
+abstract class ProjectExecTransformerBase(val list: Seq[NamedExpression], val input: SparkPlan)
```
Comment (yma11) [inline gluten-substrait/src/main/scala/org/apache/gluten/execution/BasicPhysicalOperatorTransformer.scala:181]: @zhztheplayer here is a change from case class to abstract class, any concerns about this?

**Review by yma11 — APPROVED** [review]
Thanks for contribution!

**Comment by felipepessoto** [comment]
@dcoliversun, I think I found a bug with this change and FilterMetricsUpdater. The DeleteCommand query is:

                    data.filter(new Column(cond))
                      .select(input_file_name())
                      .filter(new Column(incrDeletedCountExpr))
                      .distinct()
                      .as[String]
                      .collect()

If you remove the `.select(input_file_name())`, and replace it by something like `.select(col("_metadata.file_path"))`, or `.select(org.apache.spark.sql.functions.lit("TEST"))` the FilterMetricsUpdater.updateNativeMetrics is not called with correct `operatorMetrics.outputRows`.

I guess some shortcut is taken when we don't have a UDF. But this shouldn't happen, since IncrementMetric is Nondeterministic. I'm not sure if this is the case though, Idk how to troubleshoot this.

I added some println:

```
class FilterMetricsUpdater(
    val metrics: Map[String, SQLMetric],
    val extraMetrics: Seq[(String, SQLMetric)])
  extends MetricsUpdater {

  override def updateNativeMetrics(opMetrics: IOperatorMetrics): Unit = {
    println("FilterMetricsUpdater: 1")
    if (opMetrics != null) {
      println("FilterMetricsUpdater: 2")
      val operatorMetrics = opMetrics.asInstanceOf[OperatorMetrics]
      metrics("numOutputRows") += operatorMetrics.outputRows
      metrics("outputVectors") += operatorMetrics.outputVectors
      metrics("outputBytes") += operatorMetrics.outputBytes
      metrics("cpuCount") += operatorMetrics.cpuCount
      metrics("wallNanos") += operatorMetrics.wallNanos
      metrics("peakMemoryBytes") += operatorMetrics.peakMemoryBytes
      metrics("numMemoryAllocations") += operatorMetrics.numMemoryAllocations
      println(s"FilterMetricsUpdater: 3. ${extraMetrics.size}. $extraMetrics")
      extraMetrics.foreach {
        case (name, metric) =>
          name match {
            case "increment_metric" =>
              println(s"FilterMetricsUpdater: 4. ${operatorMetrics.outputRows}.")
              metric += operatorMetrics.outputRows
            case _ => // do nothing
          }
      }
    }
  }
}
```

This is what it shows for `.select(org.apache.spark.sql.functions.lit("TEST"))` or `.select(col("_metadata.file_path"))` and run  `build/sbt 'spark/testOnly org.apache.spark.sql.delta.DescribeDeltaHistorySuite -- -t "operation metrics - delete"'`

FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 0.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 0.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()

This is when we use the input_file_name:

FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 0.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 1.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 0.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2134, name: Some(number of rows deleted.), value: 0)))
FilterMetricsUpdater: 4. 0.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 1. List((increment_metric,SQLMetric(id: 2150, name: Some(number of rows touched), value: 0)))
FilterMetricsUpdater: 4. 3.
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()
FilterMetricsUpdater: 1
FilterMetricsUpdater: 2
FilterMetricsUpdater: 3. 0. List()

**Comment by dcoliversun** [comment]
@felipepessoto Got it. Looks like that you use new test? I will see it soon. Could you please submit an issue?

**Comment by felipepessoto** [comment]
Thanks, I opened it here: https://github.com/apache/incubator-gluten/issues/9003

**Pull request merged**